File: dev/js/Filter.js
"use strict"

class Filter {

    constructor () {}

    init ({updateFn, activation, eluAlpha}={}) {

        this.deltaWeights = this.weights.map(channel => channel.map(wRow => wRow.map(w => 0)))
        this.deltaBias = 0

        switch (updateFn) {

            case "gain":
                this.biasGain = 1
                this.weightGains = this.weights.map(channel => channel.map(wRow => wRow.map(w => 1)))
                this.getWeightGain = ([channel, row, column]) => this.weightGains[channel][row][column]
                this.setWeightGain = ([channel, row, column], v) => this.weightGains[channel][row][column] = v
                break

            case "adagrad":
            case "rmsprop":
            case "adadelta":
                this.biasCache = 0
                this.weightsCache = this.weights.map(channel => channel.map(wRow => wRow.map(w => 0)))
                this.getWeightsCache = ([channel, row, column]) => this.weightsCache[channel][row][column]
                this.setWeightsCache = ([channel, row, column], v) => this.weightsCache[channel][row][column] = v

                if (updateFn=="adadelta") {
                    this.adadeltaBiasCache = 0
                    this.adadeltaCache = this.weights.map(channel => channel.map(wRow => wRow.map(w => 0)))
                    this.getAdadeltaCache = ([channel, row, column]) => this.adadeltaCache[channel][row][column]
                    this.setAdadeltaCache = ([channel, row, column], v) => this.adadeltaCache[channel][row][column] = v
                }
                break

            case "adam":
                this.m = 0
                this.v = 0
                break
        }

        if (activation=="rrelu") {
            this.rreluSlope = Math.random() * 0.001

        } else if (activation=="elu") {
            this.eluAlpha = eluAlpha
        }
    }

    getWeight ([channel, row, column]) {
        return this.weights[channel][row][column]
    }

    setWeight ([channel, row, column], v) {
        this.weights[channel][row][column] = v
    }

    getDeltaWeight ([channel, row, column]) {
        return this.deltaWeights[channel][row][column]
    }

    setDeltaWeight ([channel, row, column], v) {
        this.deltaWeights[channel][row][column] = v
    }
}

/* istanbul ignore next */
typeof window!="undefined" && (window.Filter = Filter)
exports.Filter = Filter

File: dev/cpp/Filter.cpp

void Filter::init (int netInstance, int channels, int filterSize) {

    deltaBias = 0;
    deltaWeights = NetUtil::createVolume<double>(channels, filterSize, filterSize, 0);

    Network* net = Network::getInstance(netInstance);

    switch (net->updateFnIndex) {
        case 1: // gain
            biasGain = 1;
            weightGain = NetUtil::createVolume<double>(channels, filterSize, filterSize, 1);
            break;
        case 2: // adagrad
        case 3: // rmsprop
        case 5: // adadelta
            biasCache = 0;
            weightsCache = NetUtil::createVolume<double>(channels, filterSize, filterSize, 0);

            if (net->updateFnIndex == 5) {
                adadeltaBiasCache = 0;
                adadeltaCache = NetUtil::createVolume<double>(channels, filterSize, filterSize, 0);
            }
            break;
        case 4: // adam
            m = 0;
            v = 0;
            break;
    }

    if (net->activation == &NetMath::lrelu<Neuron>) {
        lreluSlope = net->lreluSlope;
    } else if (net->activation == &NetMath::rrelu<Neuron>) {
        rreluSlope = ((double) rand() / (RAND_MAX))/5 - 0.1;
    } else if (net->activation == &NetMath::elu<Neuron>) {
        eluAlpha = net->eluAlpha;
    }
}